### ISSUE TITLE
Codegen and Reflection handling JsonQualifier annotations differently

### ISSUE THREAD

**Issue by MarkoMilos** [square/moshi#1372]
Hi, first of all, thanks to the authors and contributors of this excellent library.

I've noticed an inconsistency between Kotlin *Codegen* and *Reflection* in handling `JsonQualifier` annotations targeting `FIELD`.

While `codegen` (generated adapter) is respecting `@JsonQualifer` annotations targeting fields `reflection` (`KotlinJsonAdapter`) won't pick them up since it is looking only for annotations targeting properties.

File/example below demonstrates the problem:

```kotlin
// 1. Define custom qualifier annotation
@JsonQualifier
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
annotation class Mask

// 2. Class using Codegen (generated adapter)
@JsonClass(generateAdapter = true)
class CodeGen(
  @Mask val password: String
)

// 3. Exact same class that uses Reflect (KotlinJsonAdapter)
class Reflect(
  @Mask val password: String
)

// 4. Custom adapter with factory for defined qualifier and String type
class MaskStringAdapter : JsonAdapter<String>() {

  override fun fromJson(reader: JsonReader): String? {
    return reader.nextString() // not important for this demo
  }

  override fun toJson(writer: JsonWriter, value: String?) {
    writer.value("*******")
  }

  class Factory : JsonAdapter.Factory {
    override fun create(type: Type, annotations: MutableSet<out Annotation>, moshi: Moshi): JsonAdapter<*>? {
      // Factory is applicable only to strings annotated with @Mask
      if (annotations.isEmpty() || type.rawType != String::class.java) return null
      val hasAnnotation = Types.nextAnnotations(annotations, Mask::class.java) != null
      return if (hasAnnotation) MaskStringAdapter() else null
    }
  }
}

fun main() {
  val moshi = Moshi.Builder()
    .add(MaskStringAdapter.Factory())
    .add(KotlinJsonAdapterFactory())
    .build()

  // 5. Serialization with generated adapter (will use MaskingAdapter for password)
  println("Codegen result:")
  println(moshi.adapter(CodeGen::class.java).toJson(CodeGen("12345")))

  // 6. Serialization with reflect adapter (won't use MaskingAdapter for password)
  println("Reflect result:")
  println(moshi.adapter(Reflect::class.java).toJson(Reflect("12345")))
}

```

Output:
```
Codegen result:
{"password":"*******"}

Reflect result:
{"password":"12345"}
```

**Expected**: both codegen and reflection adapters serialize password using MaskingAdapter
**Actual**: codegen adapter is using MaskingAdapter while reflection adapter is not
**Moshi version**: `1.12.0`

It is visible that generated adapter will query `moshi` with qualifier annotation `Mask`
```java
...

public class CodeGenJsonAdapter(
  moshi: Moshi
) : JsonAdapter<CodeGen>() {
  private val options: JsonReader.Options = JsonReader.Options.of("password")

  @field:Mask
  private val stringAtMaskAdapter: JsonAdapter<String> = moshi.adapter(String::class.java,
      Types.getFieldJsonQualifierAnnotations(javaClass, "stringAtMaskAdapter"), "password")

...
```

However it looks like in `KotlinJsonAdapter` only qualifier annotations that target `PROPERTY` are used to query for adapter: [LINK](https://github.com/square/moshi/blob/ce45a68cc35a6a6676fe31674d780e51c3f675b8/kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt#L250)

```kotlin
...

        val allAnnotations = property.annotations.toMutableList()
        var jsonAnnotation = property.findAnnotation<Json>()

        if (parameter != null) {
          allAnnotations += parameter.annotations
          if (jsonAnnotation == null) {
            jsonAnnotation = parameter.findAnnotation()
          }
        }

        val name = jsonAnnotation?.name ?: property.name
        val resolvedPropertyType = resolve(type, rawType, property.returnType.javaType)
        val adapter = moshi.adapter<Any>(
          resolvedPropertyType,
          Util.jsonAnnotations(allAnnotations.toTypedArray()),
          property.name
        )

...
```

TBH I'm not sure if this is a bug or intended behavior.

**Comment by ZacSweers** [comment]
Apologies for the delay. While this is something we could look at, is just adding `PROPERTY` not a viable solution? The ability to put property annotations on multiple locations is sort of problematic from our side, as a single property could apply the same annotation on 3-4 different declarations in bytecode.

**Comment by MarkoMilos** [comment]
In general (in most cases) it would be a viable solution for the demonstrated problem. Even removing `@Target` annotation completely would work since Kotlin would use default use-site targets in order per [documentation](https://kotlinlang.org/docs/annotations.html#annotation-use-site-targets).

I had a special case when I stumbled upon this one -> was writing a [library](https://github.com/MarkoMilos/jsonapi) that uses reflection and was expecting these annotations on the field. In tests, library implementation worked with codegen module but didn't work with reflection because of the reason documented above. I ended up writing different reflection implementations for Java vs Kotlin classes to avoid this requirement.

Even though this is a niche case and I thought that it was worth reporting/documenting :)
